### PR TITLE
Make Downloader a thin veneer over a Call.Factory.

### PR DESCRIPTION
**Pull request by JakeWharton**

### CODE REVIEW

File: picasso/src/main/java/com/squareup/picasso/BitmapHunter.java [hunk 1]
```diff
@@ -210,7 +210,7 @@ Bitmap hunt() throws IOException {
       }
     }
 
-    data.networkPolicy = retryCount == 0 ? NetworkPolicy.OFFLINE.index : networkPolicy;
+    networkPolicy = retryCount == 0 ? NetworkPolicy.OFFLINE.index : networkPolicy;
```
Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/BitmapHunter.java]: Bug!

File: picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java [hunk 2]
```diff
@@ -43,29 +42,59 @@ public NetworkRequestHandler(Downloader downloader, Stats stats) {
...
-    if (loadedFrom == NETWORK && response.getContentLength() > 0) {
-      stats.dispatchDownloadFinished(response.getContentLength());
+    if (loadedFrom == NETWORK && body.contentLength() > 0) {
+      stats.dispatchDownloadFinished(body.contentLength());
     }
+    InputStream is = body.byteStream();
     return new Result(is, loadedFrom);
   }
 
+  private okhttp3.Request createRequest(Request request, int networkPolicy) {
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java]: static?

File: picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java [hunk 3]
```diff
@@ -43,29 +42,59 @@ public NetworkRequestHandler(Downloader downloader, Stats stats) {
...
-  @Override @Nullable public Result load(Request request, int networkPolicy) throws IOException {
-    Response response = downloader.load(request.uri, request.networkPolicy);
-    if (response == null) {
-      return null;
+  @Override public Result load(Request request, int networkPolicy) throws IOException {
+    okhttp3.Request downloaderRequest = createRequest(request, networkPolicy);
+    Response response = downloader.load(downloaderRequest);
+    ResponseBody body = response.body();
+
+    if (response.code() >= 300) {
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java]: `!response.isSuccessful()`?

File: picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java [hunk 4]
```diff
@@ -43,29 +42,59 @@ public NetworkRequestHandler(Downloader downloader, Stats stats) {
...
+      body.close();
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java]: Dumb question, but should all of the `body.close()` calls here close quietly?

Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java]: That's the implementation of `ResponseBody.close()`.

Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java]: ah, yep!

**Pull request merged**